skorch/history.py

```diff
@@ -145,24 +148,68 @@ def record_batch(self, attr, value):
         batch.
 
         """
+        # pylint: disable=invalid-sequence-index
         self[-1]['batches'][-1][attr] = value
 
     def to_list(self):
         """Return history object as a list."""
         return list(self)
 
     def __getitem__(self, i):
+        # This implementation resolves indexing backwards,
+        # i.e. starting from the batches, then progressing to the
+        # epochs.
         if isinstance(i, (int, slice)):
-            return super().__getitem__(i)
-
-        x = self
-        if isinstance(i, tuple):
-            for part in i:
-                x_dirty = partial_index(x, part)
-                x = filter_missing(x_dirty)
-                # pylint: disable=unidiomatic-typecheck
-                if type(x) is _missingno:
-                    raise x.e
-            return x
-        raise ValueError("Invalid parameter type passed to index. "
-                         "Pass string, int or tuple.")
+            i = (i,)
+
+        # i_e: index epoch
+        # k_e: key epoch
+        # i_b: index batch
+        # k_e: key batch
+        i_e, k_e, i_b, k_b = _unpack_index(i)
+        keyerror_msg = "Key '{}' was not found in history."
+        batches_unpacked = False
+
+        if i_b is not None and k_e != 'batches':
+            raise KeyError("History indexing beyond the 2nd level is "
+                           "only possible if key 'batches' is used, "
+                           "found key '{}'.".format(k_e))
+
+        items = self.to_list()
+
+        # extract indices of batches
```

Comment:
I would like to have a hint of what portion is handled in each section, e.g. a comment like

```python
# extract indices of batches
# handles: items = history[..., k_e, i_b]

[...]

# extract keys of batches
# handles: items = history[..., k_e, i_b][k_b]

[...]

# extract the batches, but only if not already done 
# handles: items = history[..., k_e]

[...]

# extract the epochs 
# handles: items = history[i_e, ..., ..., ...]
```